pkg/artifacts/signable.go

```diff
@@ -436,3 +463,119 @@ func (oa *OCIArtifact) FullKey(obj interface{}) string {
+	var objs []interface{}
+	for _, obj := range extractSBOMFromResults(ctx, tektonObject) {
+		objs = append(objs, objects.NewSBOMObject(obj.SBOMURI, obj.SBOMFormat, obj.URI, obj.Digest, tektonObject))
+	}
+	return objs
+}
+
+func (sa *SBOMArtifact) StorageBackend(cfg config.Config) sets.Set[string] {
+	return cfg.Artifacts.SBOM.StorageBackend
+}
+func (sa *SBOMArtifact) Signer(cfg config.Config) string {
+	return cfg.Artifacts.SBOM.Signer
+}
+func (sa *SBOMArtifact) PayloadFormat(cfg config.Config) config.PayloadType {
+	return config.PayloadType(cfg.Artifacts.SBOM.Format)
+}
+
+func (sa *SBOMArtifact) FullKey(obj interface{}) string {
+	v := obj.(*objects.SBOMObject)
+	return v.GetSBOMURL()
+}
+
+func (sa *SBOMArtifact) ShortKey(obj interface{}) string {
+	v := obj.(*objects.SBOMObject)
+	return v.GetSBOMURL()
+}
+func (sa *SBOMArtifact) Type() string {
+	return "sbom"
+}
+func (sa *SBOMArtifact) Enabled(cfg config.Config) bool {
+	return cfg.Artifacts.SBOM.Enabled()
+}
+
+func extractSBOMFromResults(ctx context.Context, tektonObject objects.TektonObject) []*StructuredSignable {
+	logger := logging.FromContext(ctx)
+	var objs []*StructuredSignable
+
+	ss := extractTargetFromResults(ctx, tektonObject, "IMAGE_URL", "IMAGE_DIGEST", "IMAGE_SBOM_URL", "IMAGE_SBOM_FORMAT")
+	for _, s := range ss {
+		if s == nil || s.Digest == "" || s.URI == "" || s.SBOMURI == "" || s.SBOMFormat == "" {
+			continue
+		}
+		if _, err := name.NewDigest(s.SBOMURI); err != nil {
+			logger.Errorf("error getting digest for SBOM image %s: %v", s.SBOMURI, err)
+			continue
+		}
+		objs = append(objs, s)
+	}
+
+	var images []name.Digest
+	var sboms []string
+	var sbomsFormat string
+	// look for a comma separated list of images and their SBOMs
+	for _, key := range tektonObject.GetResults() {
+		switch key.Name {
+		case "IMAGES":
+			for _, img := range strings.FieldsFunc(key.Value.StringVal, split) {
+				trimmed := strings.TrimSpace(img)
+				if trimmed == "" {
+					continue
+				}
+
+				dgst, err := name.NewDigest(trimmed)
+				if err != nil {
+					logger.Errorf("error getting digest for img %s: %v", trimmed, err)
+					continue
+				}
+				images = append(images, dgst)
+			}
+		case "SBOMS":
+			for _, img := range strings.FieldsFunc(key.Value.StringVal, split) {
+				trimmed := strings.TrimSpace(img)
+				if trimmed == "" {
+					continue
+				}
+				if _, err := name.NewDigest(trimmed); err != nil {
+					logger.Errorf("error getting digest for SBOM image %s: %v", trimmed, err)
+					continue
+				}
+				sboms = append(sboms, trimmed)
+			}
+		case "SBOMS_FORMAT":
+			f := strings.TrimSpace(key.Value.StringVal)
+			if f != "" {
+				sbomsFormat = f
+			}
+		}
+	}
+
+	if len(images) != len(sboms) {
+		logger.Warnf("IMAGES and SBOMS do not contain the same amount of entries")
+		return objs
+	}
+
+	if len(sboms) > 0 && sbomsFormat == "" {
+		logger.Warnf("SBOMS_FORMAT not specified")
+		return objs
+	}
+
+	for i, sbom := range sboms {
+		img := images[i]
```

Comment:
Should we try to discover .sbom images from the base URLs instead of using params? 🤔

Reply:
> This is making the assumption that SBOMs are 1:1 with images - is there a way we can enforce that this pairing is correct?

Not unless we crack open the SBOM and add support to specific formats. For example, if it's SPDX, we can look at `.name`. If it's cyclonedx, we can look at `.metadata.component`. I'm trying to avoid having to teach Chains about any specific format.

> Should we try to discover .sbom images from the base URLs + digest instead of using params? 

I think that's worth discussing. The reason I didn't take that approach is because the relationship between an SBOM attachment and the image is weak. It's simply done by tag convention. It would be trivial for someone to trick Chains into signing an SBOM  that was created elsewhere without leaving any trace in the pipeline run. I think it's important to include the digest of the SBOM image in the SLSA Provenance so it can be properly audited and verified later on.

Thoughts?

Reply:
I've reworked this section. It is much simplified by using blobs directly.